File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_01_lists/P07.scala
package com.shekhargulati.ninetynine_problems._01_lists

/**
  * <b>(**) Flatten a nested list structure</b>
  * <p>
  * Transform a list, possibly holding lists as elements into a 'flat' list by replacing each list with its elements (recursively).
  * </p>
  */
object P07 {

  def flatten(list: List[Any]): List[Any] = {
    var result = List[Any]()
    for (el <- list) {
      el match {
        case xs: List[Any] =>
          result = result ++: flatten(xs)
        case _ =>
          result = result :+ el
      }
    }
    result
  }

  def flatten_flatMap(list: List[Any]): List[Any] = list.flatMap(el => el match {
    case xs: List[Any] => flatten_flatMap(xs)
    case _ => List(el)
  })

}

File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_01_lists/P07.java
package com.shekhargulati.ninetynine_problems._01_lists;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;

/**
 * <b>(**) Flatten a nested list structure</b>
 * <p>
 * Transform a list, possibly holding lists as elements into a 'flat' list by replacing each list with its elements (recursively).
 * </p>
 */
public class P07 {

    public static <T> List<T> flatten(List<?> list, Class<T> elementType) {
        List<T> flatten = new ArrayList<>();
        list.forEach(e -> {
            if (e instanceof List) {
                flatten.addAll(flatten((List<?>) e, elementType));
            } else {
                flatten.add((T) e);
            }
        });
        return flatten;
    }

    public static <T> List<T> flatten_stream(List<?> list, Class<T> elementType) {
        return list
                .stream()
                .flatMap(e -> e instanceof List ? flatten_stream(((List<?>) e), elementType).stream() : Stream.of(e))
                .map(e -> (T) e)
                .collect(toList());
    }
}
